### PR TITLE
feat: integrate data loader interface to import

### PR DESCRIPTION
**Pull request by bzp2010**
Please answer these questions before submitting a pull request, **or your PR will get closed**.

**Why submit this pull request?**

- [ ] Bugfix
- [x] New feature provided
- [ ] Improve performance
- [ ] Backport patches

**What changes will this PR take into?**

The code implementation of the OpenAPI import function we provided previously was coupled and could not support more import types, so we split it based on the DataLoader interface and separated the import handler and the specific import protocol implementation.

**Related issues**

a part of https://github.com/apache/apisix-dashboard/issues/2465

**Checklist:**

- [x] Did you explain what problem does this PR solve? Or what new features have been added?
- [x] Have you added corresponding test cases?
- [x] Have you modified the corresponding document?
- [x] Is this PR backward compatible? If it is not backward compatible, please discuss on the mailing list first

### CODE REVIEW

**Comment by bzp2010** [comment]
### Update

1. The changes in master have been merged and the duplicate code now no longer exists.
2. I added a default value, `0.0.0.0`, for the upstream nodes field generated during openapi3 import. If this data is not added, it will be ignored due to omitempty during json.Marshal, resulting in a schema checksum error.
3. I removed the test case imported in the front-end and it will be re-added in the next frontend PR #2480.

File: api/internal/handler/data_loader/route_import.go [hunk 1]
```diff
@@ -19,535 +19,317 @@ package data_loader
...
+			// has been found, the HTTP method is checked for overlap, and
+			// if there is overlap it is determined to be a duplicate route
+			// and the import is rejected.
+			Predicate: func(obj interface{}) bool {
+				r := obj.(*entity.Route)
+				isMethodDuplicated := len(intersect.Hash(r.Methods, route.Methods)) > 0
+
+				// Check URI and host duplication
+				// First check for duplicate URIs
+				if (r.URI != "" && route.URI != "" && r.URI == route.URI) || len(intersect.Hash(r.Uris, route.Uris)) > 0 {
```
Comment (tokers) [inline api/internal/handler/data_loader/route_import.go:156]: This if condition is complicated. Can we encapsulate a method?

**Review by starsz — APPROVED** [review]

File: api/internal/handler/data_loader/route_import.go [hunk 2]
```diff
@@ -19,535 +19,317 @@ package data_loader
...
-		}
-		if v.Post != nil {
-			value := PathValue{
-				Method: http.MethodPost,
-				Value:  v.Post,
+	for _, route := range data.Routes {
+		_, err := h.routeStore.Create(ctx, &route)
+		if err != nil {
+			if errs[store.HubKeyRoute] == nil {
+				errs[store.HubKeyRoute] = []string{}
```
Comment (tokers) [inline api/internal/handler/data_loader/route_import.go:205]: The explicit initialization is not necessary because append will do it .

**Review by tokers — COMMENTED** [review]
What's the plan for testing if imported OAS can work?

File: api/internal/handler/data_loader/route_import.go [hunk 3]
```diff
@@ -19,535 +19,317 @@ package data_loader
...
 	}
 
-	if len(swagger.Paths) < 1 {
-		return &data.SpecCodeResponse{StatusCode: http.StatusBadRequest},
-			consts.ErrImportFile
+	var l loader.Loader
+	switch LoaderType(input.Type) {
+	case LoaderTypeOpenAPI3:
+		l = &openapi3.Loader{
+			MergeMethod: input.MergeMethod == "true",
```
[the remark this answers is not in the document]
Comment (bzp2010) [reply api/internal/handler/data_loader/route_import.go:113]: Hi, @nic-6443 

The handler currently nests a library called `droplet`, and as I read through its code, it looks like the data cannot be directly parsed as a `bool` type. It only handles `string` and `[]byte` types. So I have to use this inelegant writing style here.

reference: https://github.com/ShiningRush/droplet/blob/53817015cd1b7a7bd47da219e7a72cfb71af5968/middleware/http_input.go#L124-L143

BTW, this droplet is expected to be removed in the refactoring and used in its native form `gin`.

File: api/internal/handler/data_loader/loader/openapi3/import.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+	"github.com/apisix/manager-api/internal/utils/consts"
+)
+
+func (o Loader) Import(input interface{}) (*loader.DataSets, error) {
+	if input == nil {
+		return nil, errors.New("input is nil")
+	}
+
+	// load OAS3 document
+	swagger, err := openapi3.NewSwaggerLoader().LoadSwaggerFromData(input.([]byte))
```
[the remark this answers is not in the document]
Comment (bzp2010) [reply api/internal/handler/data_loader/loader/openapi3/import.go]: Hi, @tokers 

I noticed that this section is no longer included in this PR, it currently looks like this. Where are you quoting these codes from? 🤔

https://github.com/apache/apisix-dashboard/pull/2474/files#diff-063d8afe9365313ac1728c1ce343e9e3129b596b5120e5319cd51526ab9d4cecL38-L47

File: api/internal/handler/data_loader/loader/openapi3/import_test.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+		case "test_customer/{customer_id}_PUT":
+			assert.Contains(t, route.Uris, "/customer/*")
+			assert.Contains(t, route.Methods, "PUT")
+			assert.Equal(t, "Update customer", route.Desc)
+			assert.Equal(t, entity.Status(0), route.Status)
+		case "test_customer/{customer_id}_DELETE":
+			assert.Contains(t, route.Uris, "/customer/*")
+			assert.Contains(t, route.Methods, "DELETE")
+			assert.Equal(t, "Remove customer", route.Desc)
+			assert.Equal(t, entity.Status(0), route.Status)
```
[the remark this answers is not in the document]
Comment (bzp2010) [reply api/internal/handler/data_loader/loader/openapi3/import_test.go]: Again, in the last PR #2460  I have solved the default failure issue and it has been merged, so this should not be an issue now. 🤔

File: api/internal/handler/data_loader/route_import.go (hunk 1 above)
Comment (bzp2010) [reply api/internal/handler/data_loader/route_import.go:156]: yes, let me try

File: api/internal/handler/data_loader/route_import.go (hunk 2 above)
Comment (bzp2010) [reply api/internal/handler/data_loader/route_import.go:205]: Oh, GET, learned something new. I will modify it. 😳

File: api/internal/handler/data_loader/loader/openapi3/import.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (bzp2010) [reply api/internal/handler/data_loader/loader/openapi3/import.go]: It has been fixed

File: api/internal/handler/data_loader/loader/openapi3/import_test.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+			assert.Contains(t, route.Methods, "GET", "GET")
+			assert.Equal(t, entity.Status(0), route.Status)
+		case "test_customers":
+			assert.Contains(t, route.Uris, "/customers")
+			assert.Contains(t, route.Methods, "GET")
+			assert.Equal(t, entity.Status(0), route.Status)
+		case "test_customer/{customer_id}":
+			assert.Contains(t, route.Uris, "/customer/*")
+			assert.Contains(t, route.Methods, "PUT", "DELETE")
+			assert.Equal(t, entity.Status(0), route.Status)
```
[the remark this answers is not in the document]
Comment (bzp2010) [reply api/internal/handler/data_loader/loader/openapi3/import_test.go]: It has been fixed

File: api/internal/handler/data_loader/route_import.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (nic-6443) [reply api/internal/handler/data_loader/route_import.go:113]: OK, got it.

File: api/internal/handler/data_loader/route_import.go (hunk 2 above)
Comment (bzp2010) [reply api/internal/handler/data_loader/route_import.go]: explicitly initialize are removed

File: api/internal/handler/data_loader/route_import.go (hunk 1 above)
Comment (bzp2010) [reply api/internal/handler/data_loader/route_import.go]: They are now drawn out as separate variables to be judged, and the difficulty of reading them has been reduced.

**Review by tokers — APPROVED** [review]

**Review by tao12345666333 — APPROVED** [review]
LGTM

wait for it to turn green

File: web/cypress/integration/route/import_export_route.spec.js [hunk 7]
```diff
@@ -1,208 +0,0 @@
-/*
```
Comment (nic-chen) [inline web/cypress/integration/route/import_export_route.spec.js:1]: why remove this file?

**Comment by bzp2010** [comment]
> ### Update
> ....
> 3. I removed the test case import in the frontend and it will be re-added in the next frontend PR [feat: support data loader in frontend #2480](https://github.com/apache/apisix-dashboard/pull/2480).

Hi, @nic-chen Handling tests here is bound to introduce a lot of front-end content, which is not good for PR reviews, so I removed it and added a new version in the next PR.

**Review by nic-chen — APPROVED** [review]

**Review by nic-6443 — APPROVED** [review]

**Pull request merged**